File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;
class Solution {
    
    static List<List<Integer>>graph=new ArrayList<>();
    public void bfs(int[] visit,int start,int v1,int v2,int flag){
        
        Deque<Integer>q=new ArrayDeque<>();
        visit[start]=flag;
        q.add(start);
        while(!q.isEmpty()){
            int now=q.poll();
            for(int i : graph.get(now)){
                if(visit[i]!=0)
                    continue;
                if((now==v1&&i==v2)||(now==v2&&i==v1))
                    continue;
                visit[i]=flag;
                q.add(i);
            }
        }
    }
    
    public int solution(int n, int[][] wires) {
        int answer = 1000;
        //전선들 중 하나를 끊어서 송전탑 개수가 가능한 비슷 하도록 두 전력망 나누도록 하자
        for (int i=0;i<=n;i++){
            graph.add(new ArrayList<>());
        }
        
        for(int i=0;i<wires.length;i++){
            int v1=wires[i][0];
            int v2=wires[i][1];
            graph.get(v1).add(v2);
            graph.get(v2).add(v1);
        }
        
        for(int i=0;i<wires.length;i++){
            int v1=wires[i][0];
            int v2=wires[i][1];
            int[]visit=new int[n+1];
            int temp=1;
            for(int j=1;j<=n;j++){
                if (visit[j]==0){
                    bfs(visit,j,v1,v2,temp);
                    temp+=1;
                }
            }
            int left=0;
            int right=0;
            for(int j=1;j<=n;j++){
                if(visit[j]==1){
                    left+=1;
                }
                else{
                    right+=1;
                }
            }
            answer=Math.min(answer,Math.abs(left-right));
        }
        
        return answer;
    }
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque
import sys

def bfs(start,ind,visit,graph,v1,v2):
    visit[start]=ind
    q=deque()
    q.append(start)
    while q:
        a1=q.popleft()
        for i in graph[a1]:
            if v1==a1 and v2==i:
                continue
            if v1==i and v2==a1:
                continue
            if visit[i]==0:
                visit[i]=ind
                q.append(i)
                


def solution(n, wires):
    answer = sys.maxsize
    #전력망을 2개로 분할 하려고 한다
    # 두 전력 망이 갖게 되는 송전탑의 개수를 최대한 비슷하게 맞추고자 한다
    #송전탑 개수, 전선 정보 가 매개 변수로 주어진다
    graph=[[]for _ in range(n+1)]
    for a1,a2 in wires:
        graph[a1].append(a2)
        graph[a2].append(a1)
    
    for a1,a2 in wires:
        #이제 이 두 정점을 있는 거를 위주로 작업해 보도록 하자
        visit=[0]*(n+1)
        visit[0]=-1
        ind=1
        check=0
        for i in range(1,len(visit)):
            if visit[i]==0:
                bfs(i,ind,visit,graph,a1,a2)
                ind+=1
            if ind >=4:
                check=1
                break
        if check==0:
            a1=visit.count(1)
            a2=visit.count(2)
            answer=min(answer,abs(a1-a2))
                
    
    return answer
    
    